### PR TITLE
Feature/transaction cost

### PR DESCRIPTION
**Pull request by dabasov**
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR  
  
Associated PRs (Link as appropriate):
...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/transaction-cost  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/transaction-cost  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/chaincore/chain/state.go [hunk 1]
```diff
@@ -146,6 +147,30 @@ func (c *Chain) UpdateState(ctx context.Context, b *block.Block, bState util.Mer
...
+	)
+
+	if txn.TransactionType == transaction.TxnTypeSmartContract {
+		var scData sci.SmartContractTransactionData
+		dataBytes := []byte(txn.TransactionData)
+		err := json.Unmarshal(dataBytes, &scData)
+		if err != nil {
+			logging.Logger.Error("Error while decoding the JSON from transaction",
+				zap.Any("input", txn.TransactionData), zap.Any("error", err))
+			return math.MaxInt32
```
Comment (peterlimg) [inline code/go/0chain.net/chaincore/chain/state.go]: Return an error here should be more reasonable, so as when the TransactionType is incorrect.

File: code/go/0chain.net/miner/protocol_block.go [hunk 2]
```diff
@@ -640,6 +652,12 @@ func txnIterHandlerFunc(mc *Chain,
 			logging.Logger.Error("generate block (invalid entity)", zap.Any("entity", qe))
 			return true
 		}
+		cost := mc.EstimateTransactionCost(ctx, mc.GetLatestFinalizedBlock(), mc.GetLatestFinalizedBlock().ClientState, txn)
```
Comment (peterlimg) [inline code/go/0chain.net/miner/protocol_block.go]: If there's an Unmarshal error happens in the EstimateTransactionCost, the cost will be > MaxBlockCost and cause unexpected behaviour, the EstimateTransactionCost should return error.

File: code/go/0chain.net/miner/protocol_block_main.go [hunk 3]
```diff
@@ -4,11 +4,10 @@
 package miner
 
 import (
-	"context"
-
 	"0chain.net/chaincore/block"
 	"0chain.net/chaincore/node"
 	"0chain.net/core/datastore"
+	"context"
```
Comment (peterlimg) [inline code/go/0chain.net/miner/protocol_block_main.go]: Run goimports to format

File: code/go/0chain.net/smartcontract/faucetsc/sc.go [hunk 4]
```diff
@@ -57,6 +58,24 @@ func (fc *FaucetSmartContract) GetRestPoints() map[string]smartcontractinterface
 	return fc.SmartContract.RestHandlers
 }
 
+func (fc *FaucetSmartContract) GetCost(t *transaction.Transaction, funcName string, balances c_state.StateContextI) int {
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/faucetsc/sc.go]: We should return error if any, the caller needs to know what happened and take action base on the returned value and error.

File: code/go/0chain.net/smartcontract/interestpoolsc/sc.go [hunk 5]
```diff
@@ -1,8 +1,11 @@
 package interestpoolsc
 
 import (
+	"0chain.net/core/logging"
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/interestpoolsc/sc.go]: Run goimports to format

File: code/go/0chain.net/smartcontract/interestpoolsc/sc.go [hunk 6]
```diff
@@ -59,6 +62,24 @@ func (ipsc *InterestPoolSmartContract) GetRestPoints() map[string]smartcontracti
 	return ipsc.RestHandlers
 }
 
+func (ipsc *InterestPoolSmartContract) GetCost(t *transaction.Transaction, funcName string, balances c_state.StateContextI) int {
+	node := ipsc.getGlobalNode(balances, funcName)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/interestpoolsc/sc.go]: the name of node is conflict with the package node name, please rename it.

File: code/go/0chain.net/smartcontract/minersc/sc.go [hunk 7]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"go.uber.org/zap"
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/minersc/sc.go]: Run goimports to formant

File: code/go/0chain.net/smartcontract/minersc/sc.go [hunk 8]
```diff
@@ -87,6 +89,28 @@ func (msc *MinerSmartContract) GetRestPoints() map[string]sci.SmartContractRestH
 	return msc.RestHandlers
 }
 
+func (msc *MinerSmartContract) GetCost(t *transaction.Transaction, funcName string, balances cstate.StateContextI) int {
+	node, err := getGlobalNode(balances)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/minersc/sc.go]: Rename node

File: code/go/0chain.net/smartcontract/minersc/sc.go [hunk 9]
```diff
@@ -87,6 +89,28 @@ func (msc *MinerSmartContract) GetRestPoints() map[string]sci.SmartContractRestH
...
+		Logger.Error("can't get global node")
+		return math.MaxInt32
+	}
+	if node.Cost == nil {
+		Logger.Error("can't get cost")
+		return math.MaxInt32
+	}
+	cost, ok := node.Cost[funcName]
+	if !ok {
+		for f, _ := range node.Cost {
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/minersc/sc.go]: can be:
for f := range node.Cost {

File: code/go/0chain.net/smartcontract/minersc/settings.go [hunk 10]
```diff
@@ -207,6 +210,7 @@ func (gn *GlobalNode) set(key string, change string) error {
 			return fmt.Errorf("%s must be a hes string: %v", key, err)
 		}
 		gn.setKey(key, change)
+	case smartcontract.Cost:
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/minersc/settings.go:213]: Not implemented yet?

File: code/go/0chain.net/smartcontract/storagesc/sc.go [hunk 11]
```diff
@@ -1,8 +1,11 @@
 package storagesc
 
 import (
+	"0chain.net/core/logging"
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/sc.go]: Run goimports to format

File: code/go/0chain.net/smartcontract/vestingsc/config.go [hunk 12]
```diff
@@ -132,6 +135,8 @@ func (conf *config) update(changes *smartcontract.StringMap) error {
 			} else {
 				conf.OwnerId = value
 			}
+		case Settings[Cost]:
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/vestingsc/config.go:139]: Not implemented

File: code/go/0chain.net/smartcontract/vestingsc/sc.go [hunk 13]
```diff
@@ -2,8 +2,11 @@ package vestingsc
 
 import (
 	"0chain.net/chaincore/smartcontract"
+	"0chain.net/core/logging"
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/vestingsc/sc.go]: Run goimports to format

File: code/go/0chain.net/smartcontract/zcnsc/sc.go [hunk 14]
```diff
@@ -1,8 +1,11 @@
 package zcnsc
 
 import (
+	"0chain.net/core/logging"
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/zcnsc/sc.go]: format

File: code/go/0chain.net/smartcontract/config.go [hunk 15]
```diff
@@ -21,11 +21,12 @@ const (
 	String
 	StateBalance
 	Key
+	Cost
 	NumberOfTypes
 )
 
 var ConfigTypeName = []string{
-	"int", "int64", "int32", "time.duration", "float64", "bool", "string", "state.Balance",
+	"int", "int64", "int32", "time.duration", "float64", "bool", "string", "state.Balance", "Cost",
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/config.go]: Use lower case of "cost" to keep the consistency

File: docker.local/config/sc.yaml [hunk 16]
```diff
@@ -48,11 +61,32 @@ smart_contracts:
...
-
+    cost:
+      add_miner: 100
+      add_sharder: 100
+      delete_miner: 100
+      delete_sharder: 100
+      miner_health_check: 100
+      sharder_health_check: 100
+      contributeMpk: 100
+      shareSignsOrShares: 100
```
Comment (peterlimg) [inline docker.local/config/sc.yaml:78]: contributeMpk and shareSignsOrShares should be much higher than others, there are lots of signature verification there. We can check the smartcontracts benchmark result to set this appropriately.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: docker.local/config/sc.yaml (hunk 16 above)
Comment (dabasov) [reply docker.local/config/sc.yaml:78]: @peterlimg for sure, I think that I can use your latest performance results to find weights

File: code/go/0chain.net/chaincore/chain/state.go (hunk 1 above)
Comment (dabasov) [reply code/go/0chain.net/chaincore/chain/state.go]: well, MaxInt is more like special value for a disabled transaction, the logic is simple, this transaction will be excluded from the block. We handled error already, while writing log, and now we have more common way to handle this situation. We already know how to handle this exception upper in the stack, without extending logic with   `if err != nil {.}` further, can reduce this boilerplate check, which can make sense

File: code/go/0chain.net/miner/protocol_block.go (hunk 2 above)
Comment (dabasov) [reply code/go/0chain.net/miner/protocol_block.go]: yeah, this is correct, overflow is possible

File: code/go/0chain.net/chaincore/chain/state.go (hunk 1 above)
Comment (dabasov) [reply code/go/0chain.net/chaincore/chain/state.go]: fixed it

File: code/go/0chain.net/smartcontract/interestpoolsc/sc.go (hunk 6 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/interestpoolsc/sc.go]: fixed

File: code/go/0chain.net/smartcontract/minersc/sc.go (hunk 8 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/minersc/sc.go]: fixed

File: code/go/0chain.net/smartcontract/minersc/sc.go (hunk 9 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/minersc/sc.go]: fixed

File: code/go/0chain.net/smartcontract/minersc/settings.go (hunk 10 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/minersc/settings.go:213]: yeah, cost update is not implemented yet

Comment (dabasov) [reply code/go/0chain.net/smartcontract/minersc/settings.go:213]: I was going to delegate it to someone

File: code/go/0chain.net/smartcontract/vestingsc/config.go (hunk 12 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/vestingsc/config.go:139]: yep, wait to be delegated

File: code/go/0chain.net/smartcontract/config.go (hunk 15 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/config.go]: is it the same as Balance and should be started from capital letter?

File: code/go/0chain.net/smartcontract/interestpoolsc/sc.go (hunk 5 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/interestpoolsc/sc.go]: fixed

File: code/go/0chain.net/miner/protocol_block_main.go (hunk 3 above)
Comment (dabasov) [reply code/go/0chain.net/miner/protocol_block_main.go]: fixed

File: code/go/0chain.net/miner/protocol_block.go (hunk 2 above)
Comment (dabasov) [reply code/go/0chain.net/miner/protocol_block.go]: fixed

File: code/go/0chain.net/smartcontract/minersc/sc.go (hunk 7 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/minersc/sc.go]: fixed

File: code/go/0chain.net/smartcontract/vestingsc/config.go (hunk 12 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/vestingsc/config.go:138]: fixed

File: code/go/0chain.net/smartcontract/storagesc/sc.go (hunk 11 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/storagesc/sc.go]: fixed

File: code/go/0chain.net/smartcontract/vestingsc/sc.go (hunk 13 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/vestingsc/sc.go]: fixed

File: code/go/0chain.net/smartcontract/zcnsc/sc.go (hunk 14 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/zcnsc/sc.go]: fixed

**Review by peterlimg — DISMISSED** [review]
lgtm

**Review by peterlimg — APPROVED** [review]

**Pull request merged**